actors/miner/src/deadline_state.rs

```diff
@@ -99,6 +103,127 @@ impl Deadlines {
+                store,
+                PARTITION_EXPIRATION_AMT_BITWIDTH,
+            );
+            let mut to_early_terminations = Array::<BitField, BS>::new_with_bit_width(
+                store,
+                PARTITION_EARLY_TERMINATION_ARRAY_AMT_BITWIDTH,
+            );
+
+            // ajust epoch info of expirations_epochs and early_terminated within `Partition`
+            from_expirations_epochs.for_each(|from_epoch, expire_set| {
+                to_expirations_epochs.set(
+                    to_quant.quantize_up(from_epoch as ChainEpoch) as u64,
+                    expire_set.clone(),
+                )?;
+                Ok(())
+            })?;
+            from_early_terminations.for_each(|from_epoch, bitfield| {
+                to_early_terminations
+                    .set(to_quant.quantize_up(from_epoch as ChainEpoch) as u64, bitfield.clone())?;
+                Ok(())
+            })?;
+            moving_partition.expirations_epochs = to_expirations_epochs.flush()?;
+            moving_partition.early_terminated = to_early_terminations.flush()?;
+
+            let all_sectors = moving_partition.sectors.len();
+            let live_sectors = moving_partition.live_sectors().len();
+            let early_terminations = from_deadline.early_terminations.get(from_partition_idx);
+            if early_terminations != (from_early_terminations.count() > 0) {
+                return Err(actor_error!(illegal_state, "Deadline.early_terminations doesn't agree with Partition.early_terminated for partition {}", from_partition_idx))?;
+            }
+
+            // start updating from/to `Deadline` here
+
+            from_deadline.total_sectors -= all_sectors;
+            from_deadline.live_sectors -= live_sectors;
+            from_deadline.faulty_power -= &moving_partition.faulty_power;
+
+            to_deadline.total_sectors += all_sectors;
+            to_deadline.live_sectors += live_sectors;
+            to_deadline.faulty_power += &moving_partition.faulty_power;
+
+            // update early_terminations BitField of `Deadline`
+            if early_terminations {
+                from_deadline.early_terminations.unset(from_partition_idx);
+                to_deadline.early_terminations.set(to_partition_idx);
+            }
+
+            from_partitions.set(from_partition_idx, Partition::new(store)?)?;
+            to_partitions.set(to_partition_idx, moving_partition)?;
+        }
+
+        // update expirations_epochs Cid of Deadline.
+        {
+            let mut epochs_to_remove = Vec::<u64>::new();
+            let mut from_expirations_epochs: Array<BitField, _> =
+                Array::load(&from_deadline.expirations_epochs, store)?;
+            let mut to_expirations_epochs: Array<BitField, _> =
+                Array::load(&to_deadline.expirations_epochs, store)?;
+            from_expirations_epochs.for_each_mut(|from_epoch, from_bitfield| {
```

Comment:
Iterating the deadline's whole expiration queue is fairly expensive. I don't think we need to.
- For the from deadline, it's ok to just leave the entries there. The empty partition will no-op when the entries expire.
- For the to deadline, you can instead iterate the new _partition's_ expiration_epochs, each of which must be set in the destination deadline. 
This will iterate only the moved partition's queue, not any whole deadline's queue. It will be easier to fit this inside `Deadline::add/remove_partition` since you only need to access one deadline at a time. Again easier testing.

Note that this approach is only ok on the precondition that there are no faulty sectors, which would appear twice in the source deadline and must be propagated at both epochs. Document this.

Reply:
But it's essentially the same expensive as [`expiration_epochs.cut`](https://github.com/filecoin-project/builtin-actors/blob/0679eecb12bf2ec7091866c442f410062b21bf68/actors/miner/src/deadline_state.rs#L728) which is called in [`remove_partitions`](https://github.com/filecoin-project/builtin-actors/blob/0679eecb12bf2ec7091866c442f410062b21bf68/actors/miner/src/deadline_state.rs#L620), do we really want to save some gas by leaving some garbage data?